File: sdk/ts/src/noise/session.ts
import { CipherState } from './cipherState.js';
import { HandshakeState } from './handshakeState.js';
import type { KeyPair } from './primitives.js';
import { NotReadyError } from '../errors.js';

/** E2E envelope types (docs/PROTOCOL.md §6): binary WebSocket frames are one byte of
 * envelope type followed by the Noise message or transport ciphertext. */
export const ENVELOPE_HANDSHAKE = 0x01;
export const ENVELOPE_TRANSPORT = 0x02;

export function encodeEnvelope(kind: number, payload: Uint8Array): Uint8Array {
  const out = new Uint8Array(1 + payload.length);
  out[0] = kind;
  out.set(payload, 1);
  return out;
}

export function decodeEnvelope(frame: Uint8Array): { kind: number; payload: Uint8Array } | null {
  if (frame.length < 1) return null;
  return { kind: frame[0]!, payload: frame.slice(1) };
}

type SessionStatus = 'handshaking' | 'ready' | 'failed';

/**
 * Drives one Noise XX handshake (docs/PROTOCOL.md §6) as either role and, once it
 * completes, encrypts/decrypts transport messages for that session. Does not itself
 * implement the make-before-break replacement policy — see peer.ts for the wrapper
 * that decides when a new NoiseSession may replace an existing one.
 */
export class NoiseSession {
  private status: SessionStatus = 'handshaking';
  private readonly handshake: HandshakeState;
  private gotFirstMessage = false;
  private sendCipher?: CipherState;
  private recvCipher?: CipherState;
  private peerId?: string | undefined;
  private _peerStaticKey?: Uint8Array | undefined;

  private readonly readyPromise: Promise<void>;
  private resolveReady!: () => void;
  private rejectReady!: (err: Error) => void;

  private constructor(private readonly initiator: boolean, staticKeyPair: KeyPair) {
    this.handshake = new HandshakeState(initiator, staticKeyPair);
    this.readyPromise = new Promise((resolve, reject) => {
      this.resolveReady = resolve;
      this.rejectReady = reject;
    });
    this.readyPromise.catch(() => {}); // never an unhandled rejection if nobody awaits
  }

  /** Starts a handshake as the Noise initiator. Returns the msg1 payload to send. */
  static startAsInitiator(staticKeyPair: KeyPair): { session: NoiseSession; msg1: Uint8Array } {
    const session = new NoiseSession(true, staticKeyPair);
    const { message } = session.handshake.writeMessage();
    return { session, msg1: message };
  }

  /** Starts a handshake as the Noise responder, ready to receive msg1. */
  static startAsResponder(staticKeyPair: KeyPair): NoiseSession {
    return new NoiseSession(false, staticKeyPair);
  }

  get peerStaticKey(): Uint8Array | undefined {
    return this._peerStaticKey;
  }

  isReady(): boolean {
    return this.status === 'ready';
  }

  /** True once the handshake has finished, successfully or not. */
  isSettled(): boolean {
    return this.status !== 'handshaking';
  }

  /** Resolves once the handshake finishes; rejects if it fails. */
  waitReady(): Promise<void> {
    return this.readyPromise;
  }

  /**
   * Feeds one received handshake message. Returns a reply to send, if any. A bad
   * peer message fails this session (observable via waitReady()); it does not throw.
   */
  handleHandshakeMessage(data: Uint8Array): Uint8Array | undefined {
    if (this.status !== 'handshaking') {
      throw new Error('relayly: handshake message received after handshake finished');
    }

    try {
      if (this.initiator) {
        // The only message an initiator ever receives is msg2; writing msg3
        // completes the handshake for the initiator (matches HandshakeState).
        this.handshake.readMessage(data);
        const { message, result } = this.handshake.writeMessage();
        if (result) {
          this.finish(result.cipherState1, result.cipherState2);
        }
        return message;
      }

      if (!this.gotFirstMessage) {
        this.gotFirstMessage = true;
        this.handshake.readMessage(data); // msg1
        const { message } = this.handshake.writeMessage(); // msg2
        return message;
      }

      const { result } = this.handshake.readMessage(data); // msg3
      if (result) {
        this.finish(result.cipherState2, result.cipherState1);
      }
      return undefined;
    } catch (err) {
      this.fail(err instanceof Error ? err : new Error(String(err)));
      return undefined;
    }
  }

  private finish(sendCipher: CipherState, recvCipher: CipherState): void {
    this.sendCipher = sendCipher;
    this.recvCipher = recvCipher;
    this._peerStaticKey = this.handshake.peerStaticKey;
    this.status = 'ready';
    this.resolveReady();
  }

  private fail(err: Error): void {
    if (this.status === 'failed') return;
    this.status = 'failed';
    this.rejectReady(err);
  }

  encrypt(plaintext: Uint8Array): Uint8Array {
    if (this.status !== 'ready' || !this.sendCipher) {
      throw new NotReadyError(this.peerId ?? '?');
    }
    return this.sendCipher.encryptWithAd(new Uint8Array(0), plaintext);
  }

  decrypt(ciphertext: Uint8Array): Uint8Array {
    if (this.status !== 'ready' || !this.recvCipher) {
      throw new NotReadyError(this.peerId ?? '?');
    }
    return this.recvCipher.decryptWithAd(new Uint8Array(0), ciphertext);
  }

  /** Used only to make NotReadyError's message more useful; purely cosmetic. */
  setPeerIdHint(peerId: string): void {
    this.peerId = peerId;
  }
}

File: sdk/cpp/src/noise/session.cpp
#include "noise/session.hpp"

#include <stdexcept>

namespace relayly::noise {

std::vector<std::uint8_t> EncodeEnvelope(std::uint8_t kind, const std::vector<std::uint8_t>& payload) {
  std::vector<std::uint8_t> frame;
  frame.reserve(payload.size() + 1);
  frame.push_back(kind);
  frame.insert(frame.end(), payload.begin(), payload.end());
  return frame;
}

std::optional<std::pair<std::uint8_t, std::vector<std::uint8_t>>> DecodeEnvelope(
    const std::vector<std::uint8_t>& frame) {
  if (frame.empty()) return std::nullopt;
  return std::make_pair(frame.front(), std::vector<std::uint8_t>(frame.begin() + 1, frame.end()));
}

std::pair<NoiseSession, std::vector<std::uint8_t>> NoiseSession::AsInitiator(const KeyPair& static_keypair) {
  NoiseSession session;
  session.handshake_.emplace(true, static_keypair);
  auto result = session.handshake_->WriteMessage();
  // A single WriteMessage can never complete the 3-message XX pattern on its own.
  return {std::move(session), std::move(result.message)};
}

NoiseSession NoiseSession::AsResponder(const KeyPair& static_keypair) {
  NoiseSession session;
  session.handshake_.emplace(false, static_keypair);
  return session;
}

void NoiseSession::Finish(const HandshakeState::SplitResult& split) {
  send_ = split.send;
  recv_ = split.recv;
  peer_static_ = split.peer_static;
  status_ = SessionStatus::kReady;
  handshake_.reset();
}

NoiseSession::HandshakeResult NoiseSession::HandleHandshakeMessage(const std::vector<std::uint8_t>& data) {
  if (status_ != SessionStatus::kHandshaking || !handshake_.has_value()) {
    status_ = SessionStatus::kFailed;
    return {std::nullopt, true};
  }

  try {
    auto read_result = handshake_->ReadMessage(data);
    if (read_result.split.has_value()) {
      Finish(*read_result.split);
      return {std::nullopt, true};
    }

    auto write_result = handshake_->WriteMessage();
    if (write_result.split.has_value()) {
      auto reply = write_result.message;
      Finish(*write_result.split);
      return {std::move(reply), true};
    }
    return {std::move(write_result.message), false};
  } catch (const std::exception&) {
    status_ = SessionStatus::kFailed;
    return {std::nullopt, true};
  }
}

const Key& NoiseSession::peer_static_key() const {
  if (status_ != SessionStatus::kReady) {
    throw NotReadyError();
  }
  return peer_static_;
}

std::vector<std::uint8_t> NoiseSession::Encrypt(const std::vector<std::uint8_t>& plaintext) {
  if (status_ != SessionStatus::kReady) {
    throw NotReadyError();
  }
  return send_->EncryptWithAd({}, plaintext);
}

std::vector<std::uint8_t> NoiseSession::Decrypt(const std::vector<std::uint8_t>& ciphertext) {
  if (status_ != SessionStatus::kReady) {
    throw NotReadyError();
  }
  return recv_->DecryptWithAd({}, ciphertext);
}

}  // namespace relayly::noise
